File: hsu_chris_reactive_methods/scripts/reactive_gap_follow.py
#!/usr/bin/env python
from __future__ import print_function
import sys
import math
import numpy as np

#ROS Imports
import rospy
from sensor_msgs.msg import Image, LaserScan
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive

class reactive_follow_gap:
    def __init__(self):
        #Topics & Subscriptions,Publishers
        lidarscan_topic = '/scan'
        drive_topic = '/nav'
        self.lidar_sub = rospy.Subscriber(lidarscan_topic, LaserScan, self.lidar_callback)
        self.drive_pub = rospy.Publisher(drive_topic, AckermannDriveStamped, queue_size=10)
    
    @staticmethod
    def running_mean(x,N):
        cumsum = np.cumsum(np.insert(x,0,0))
        return (cumsum[N:]- cumsum[:-N]) / float(N)

    def preprocess_lidar(self, data, ranges, angles):
        """ Preprocess the LiDAR scan array. Expert implementation includes:
            1.Setting each value to the mean over some window
            2.Rejecting high values (eg. > 3m)
        """
        N = 5 #running mean window
        valid = np.arange(270,810)  #real lidar is from -1.57 to 1.57
        proc_ranges = ranges[valid]
        proc_angles = angles[valid]
        proc_ranges = self.running_mean(proc_ranges,N)
 
        return proc_ranges, proc_angles

    def bubble_lidar(self, ranges, angles):
        radius = 0.2
        #Find closest point (point of interest) to LiDAR
        poi_val = np.min(ranges)
        poi_idx = np.argmin(ranges)
        #Eliminate all points inside 'bubble' (set them to zero) 
        if poi_val == 0:
            alpha = np.arcsin(0)
        elif poi_val > radius:
            alpha = np.arcsin(radius/poi_val)
        else:
            alpha = np.arcsin(1)
        bub_idx_r = np.argmin(np.abs(angles-(angles[poi_idx]-alpha)))
        bub_idx_l = np.argmin(np.abs(angles-(angles[poi_idx]+alpha)))
        bubble = np.arange(bub_idx_r,bub_idx_l)

        bubble_ = np.clip(bubble,0,len(ranges)-1)
        ranges[bubble_] = 0.0

        return ranges, poi_val, poi_idx

    def find_max_gap(self, ranges):
        """ Return the start index & end index of the max gap in ranges
        """
        ranges[ranges!=0] = 1.0
        ranges1 = np.hstack((np.copy(ranges),0))
        ranges2 = np.hstack((0,np.copy(ranges)))
        check = ranges1 - ranges2
        #start and end indices of gaps
        start = np.where(check==1)[0]
        end = np.where(check==-1)[0]-1
        #check which gap is larger
        big_gap_idx = np.argmax(end-start)
        return start[big_gap_idx], end[big_gap_idx]
    
    def find_best_point(self, start_i, end_i, ranges):
        """Start_i & end_i are start and end indicies of max-gap range, respectively
        Return index of best point in ranges
	Naive: Choose the furthest point within ranges and go there
        """
        # best_idx = np.argmax(ranges[start_i:end_i])
        best_idx = int(np.mean([start_i,end_i]))
        return best_idx

    def lidar_callback(self, data):
        """ Process each LiDAR scan as per the Follow Gap algorithm & publish an AckermannDriveStamped Message
        """
        angles = np.linspace(data.angle_min,data.angle_max,len(data.ranges))
        ranges = np.array(data.ranges)
        proc_ranges, proc_angles = self.preprocess_lidar(data, ranges, angles)
        # Make a copy for analysis
        virt_ranges = np.copy(proc_ranges)
        # Zero out points in bubble
        virt_ranges, poi_val, poi_idx = self.bubble_lidar(virt_ranges, proc_angles)
        #Find max length gap in virtual range
        start, end = self.find_max_gap(virt_ranges)
        #Find the best point in the gap of processed ranges
        best = self.find_best_point(start, end, proc_ranges)
        #Associated angle with best point
        angle = proc_angles[best]
        abs_angle = np.absolute(angle)

        if 0 <= abs_angle <= (10*np.pi/180):
            speed = 1.8
        elif (10*np.pi/180) < abs_angle < (20*np.pi/180):
            speed = 1.2
        else:
            speed = 1.0

        # Publish drive message
        drive_msg = AckermannDriveStamped()
        drive_msg.header.stamp = rospy.Time.now()
        drive_msg.header.frame_id = "laser"
        drive_msg.drive.steering_angle = angle
        drive_msg.drive.speed = speed
        self.drive_pub.publish(drive_msg)

def main(args):
    rospy.init_node("gap_follow_node", anonymous=True)
    rfgs = reactive_follow_gap()
    rospy.sleep(0.1)
    rospy.spin()

if __name__ == '__main__':
    main(sys.argv)

File: hsu_chris_reactive_methods/src/reactive_gap_follow.cpp
#include <ros/ros.h>
// #include <iostream>
// #include <nav_msgs/Odometry.h>
#include <sensor_msgs/LaserScan.h>
#include <ackermann_msgs/AckermannDriveStamped.h>

#include <tf/transform_datatypes.h>
#include <vector>
// TODO: include ROS msg type headers and libraries

class ReactiveFollowGap{
// The class follows the largest gap
private:
    int const BUBBLE_SZ = 10;
    // TODO: create ROS subscribers and publishers
    ros::NodeHandle n;
    ros::Subscriber lidar_sub;
    ros::Publisher drive_pub;


public:
    struct gap {
            int start_i;
            int end_i;
    };
    struct action {
        double speed;
        double angle;
    };
    struct lidar {
        std::vector<float> ranges;
        std::vector<float> angles;
    };  
    ReactiveFollowGap() {
        n = ros::NodeHandle();
        lidar_sub = n.subscribe("/scan", 10, &ReactiveFollowGap::lidar_callback, this);
        drive_pub = n.advertise<ackermann_msgs::AckermannDriveStamped>("/nav", 10);  
    }

    lidar preprocess_lidar(const sensor_msgs::LaserScan::ConstPtr &scan_msg) {
        std::cout << scan_msg->angle_increment;
        lidar data;
        std::cout << scan_msg->angle_increment;
        float scan_angle = scan_msg->angle_min;

        // data.ranges = scan_msg->ranges;

        int num_scans = (scan_msg->angle_max-scan_msg->angle_min)/(scan_msg->angle_increment)+1;
        std::cout << num_scans;
        for (int ii = 0; ii < num_scans; ii++) 
        {
            scan_angle += scan_msg->angle_increment;
            data.angles[ii] = scan_angle;
            data.ranges[ii] = scan_msg->ranges[ii];
        }
        std::cout << data.angles[20];
        return data;
    }

    int closest_point(const sensor_msgs::LaserScan::ConstPtr &scan_msg, int &num_scans) {
        // Find closest point to lidar
        double poi = 101.0; //point of interest, closest point
        int idx;
        for (int ii = 0; ii < num_scans; ii++)
        {
        	if (scan_msg->ranges[ii] < poi)
        	{
        		poi = scan_msg->ranges[ii];
        		idx = ii;
        	}
       	}
       	return idx;
    }

    std::vector<float> bubble(const sensor_msgs::LaserScan::ConstPtr &scan_msg, int &poi_idx) {
        // Zero out all ranges in the bubble radius
        std::vector<float> virt_lidar = scan_msg->ranges;
        for (int ii = 0; ii < BUBBLE_SZ; ii++)
        {
            virt_lidar[poi_idx-ii] = 0.0;
            virt_lidar[poi_idx+ii] = 0.0;
        } 
        // for (float n : virt_lidar){
        //     std::cout << n<< '\n';
        // }
        // std::cout << "rand"<<virt_lidar[234]<< '\n'<<"min"<< virt_lidar[poi_idx]<<'\n';
        return virt_lidar;
    }

    gap find_max_gap(std::vector<float> &virt_lidar, int &num_scans) {
        // Start_i & end_i are start and end indicies of max-gap range, respectively
        gap bigGap;
        int start_temp;
        int largest = 0;
        int count = 0;
        for (int ii = 0; ii < num_scans; ii++)
        {
            if ((virt_lidar[ii] != 0) && (ii == 0))
            {
                start_temp = ii;
                count++;
            }
            else if ((virt_lidar[ii-1] == 0) && (virt_lidar[ii] != 0))
            {
                start_temp = ii;
                count++;
            }
            else if ((virt_lidar[ii-1] != 0) && (virt_lidar[ii] != 0))
            {
                count++;
            }
            else if ((virt_lidar[ii-1] != 0) && (virt_lidar[ii] == 0))
            {
                if (count > largest)
                {
                    bigGap.start_i = start_temp;
                    bigGap.end_i = ii-1;
                    largest = count;
                    count = 0;
                }
                else
                {
                    count = 0;
                }
            }
            else 
            {
                count = 0;
            }
        }
        // for (float n : virt_lidar){
        //     std::cout << n<< '\n';
        // }
        // std::cout << "start_i"<< start_i << '\n';
        // std::cout << "end_i" << end_i << '\n';
        return bigGap;
    }

    int find_best_point(std::vector<float> &virt_lidar, gap &bigGap) {
        // Return the index of the best point in ranges
        int best_idx;
        float best_range = 0;
        for (int ii = bigGap.start_i; ii < bigGap.end_i+1; ii++)
        {
            if (virt_lidar[ii] > best_range)
            {
                best_idx = ii;
                best_range = virt_lidar[ii];
            }
        }
        return best_idx;
    }

    action drive_action(const lidar &data, int &best_idx){
        // Determine speed and angle to execute
        action drive;
        std::cout << data.angles[best_idx]<< '/n';
        // drive.angle = data.angles[best_idx];
        // // if (drive.angle)
        // drive.speed = 1.0;

        return drive;
    }

    void lidar_callback(const sensor_msgs::LaserScan::ConstPtr &scan_msg) {
        // Follow the gap algorithm
        int num_scans = (scan_msg->angle_max-scan_msg->angle_min)/(scan_msg->angle_increment)+1;
        lidar data;
        gap bigGap;
        action drive;

        // Preprocess scan_msg ranges and angles
        data = ReactiveFollowGap::preprocess_lidar(scan_msg);
        // std::cout << data.angles[20] << '\n';
        // Find closest point to LiDAR
        int poi_idx = ReactiveFollowGap::closest_point(scan_msg, num_scans);
        // Eliminate all points inside 'bubble' (set them to zero) 
        std::vector<float> virt_lidar = ReactiveFollowGap::bubble(scan_msg, poi_idx);
        // Find max length gap indexes
        bigGap = ReactiveFollowGap::find_max_gap(virt_lidar, num_scans);
        // Find the index of best (furthest) point in the gap 
        int best_idx = ReactiveFollowGap::find_best_point(virt_lidar, bigGap);
        // Determine speed and angle
        // drive = ReactiveFollowGap::drive_action(data, best_idx);
        // Publish drive message
        ackermann_msgs::AckermannDriveStamped drive_msg;
        drive_msg.header.stamp = ros::Time::now();
        drive_msg.header.frame_id = "laser";
        drive_msg.drive.steering_angle = drive.angle;
        drive_msg.drive.speed = drive.speed;
        drive_pub.publish(drive_msg);
    }

};
int main(int argc, char ** argv)
{
    ros::init(argc, argv, "gap_follow_node");
    ReactiveFollowGap rfg;
    ros::spin();
    return 0;
}